lib/bugsnag/session_tracker.rb

```diff
@@ -41,117 +36,93 @@ def create_session
         }
       }
       SessionTracker.set_current_session(new_session)
-      add_thread = Thread.new { add_session(start_time) }
-      add_thread.join()
+      add_session(start_time)
     end
 
-    def send_sessions
-      @mutex.lock
-      begin
-        deliver_sessions
-      ensure
-        @mutex.unlock
-      end
-    end
-
-    private
-    def add_session(min)
-      @mutex.lock
-      begin
-        @registered_at_exit = false unless defined?(@registered_at_exit)
-        if !@registered_at_exit
-          @registered_at_exit = true
-          at_exit do
-            if !@deliver_fallback.nil? && @deliver_fallback.status == 'sleep'
-              @deliver_fallback.terminate
-            end
-            deliver_sessions
-          end
-        end
-        @session_counts[min] ||= 0
-        @session_counts[min] += 1
-        if Time.now() - @last_sent > TIME_THRESHOLD
-          deliver_sessions
-        end
-      ensure
-        @mutex.unlock
-      end
-    end
+    alias :create_session :start_session
 
-    def deliver_sessions
-      return unless @config.track_sessions
+    def send_sessions
       sessions = []
-      @session_counts.each do |min, count|
+      counts = @session_counts.dup
+      @session_counts = Concurrent::Hash.new(0)
+      counts.each do |min, count|
         sessions << {
           :startedAt => min,
           :sessionsStarted => count
         }
-        if sessions.size >= MAXIMUM_SESSION_COUNT
-          deliver(sessions)
-          sessions = []
-        end
       end
-      @session_counts = {}
-      reset_delivery_thread
       deliver(sessions)
     end
 
-    def reset_delivery_thread
-      if !@deliver_fallback.nil? && @deliver_fallback.status == 'sleep'
-        @deliver_fallback.terminate
-      end
-      @deliver_fallback = Thread.new do
-        sleep(FALLBACK_TIME)
-        deliver_sessions
+    def start_delivery_thread
+      MUTEX.synchronize do
+        @started = nil unless defined?(@started)
+        return if @started == Process.pid
+        @started = Process.pid
+        at_exit do
+          if !@delivery_thread.nil?
+            @delivery_thread.execute
+            @delivery_thread.shutdown
+          else
+            if @session_counts.size > 0
+              send_sessions
+            end
+          end
+        end
+        @delivery_thread = Concurrent::TimerTask.new(execution_interval: 30) do
+          if @session_counts.size > 0
+            send_sessions
+          end
+        end
       end
     end
 
-    def deliver(sessionCounts)
-      if sessionCounts.length == 0
-        @config.debug("No sessions to deliver")
-        return
-      end
-      
-      if !@config.valid_api_key?
-        @config.debug("Not delivering sessions due to an invalid api_key")
+    private
+    def add_session(min)
+      @session_counts[min] += 1
+    end
+
+    def deliver(session_counts)
+      if session_counts.length == 0
```

Comment:
be consistent about using `@session_counts` or `session_counts` in this file. I would use the member variable.